src/main/java/net/iponweb/disthene/service/aggregate/AggregateService.java

```diff
@@ -0,0 +1,179 @@
+
+    private Logger logger = Logger.getLogger(AggregateService.class);
+
+    private MBassador<DistheneEvent> bus;
+    private DistheneConfiguration distheneConfiguration;
+    private Rollup baseRollup;
+
+    private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1, new NamedThreadFactory(SCHEDULER_NAME));
+
+    private final ConcurrentNavigableMap<Long, ConcurrentMap<MetricKey, AverageRecord>> accumulator = new ConcurrentSkipListMap<>();
+
+    public AggregateService(MBassador<DistheneEvent> bus, DistheneConfiguration distheneConfiguration) {
+        this.distheneConfiguration = distheneConfiguration;
+        this.baseRollup = distheneConfiguration.getCarbon().getBaseRollup();
+        this.bus = bus;
+        bus.subscribe(this);
+
+        scheduler.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                flush();
+            }
+        }, 60 - ((System.currentTimeMillis() / 1000L) % 60), RATE, TimeUnit.SECONDS);
+    }
+
+    @Handler(rejectSubtypes = false)
+    public void handle(MetricAggregateEvent metricAggregateEvent) {
+        aggregate(metricAggregateEvent.getMetric());
+    }
+
+    private ConcurrentMap<MetricKey, AverageRecord> getTimestampMap(long timestamp) {
+        ConcurrentMap<MetricKey, AverageRecord> timestampMap = accumulator.get(timestamp);
+        if (timestampMap == null) {
+            ConcurrentMap<MetricKey, AverageRecord> newTimestampMap = new ConcurrentHashMap<>();
+            timestampMap = accumulator.putIfAbsent(timestamp, newTimestampMap);
+            if (timestampMap == null) {
+                timestampMap = newTimestampMap;
+            }
+        }
+
+        return timestampMap;
+    }
+
+    private AverageRecord getAverageRecord(ConcurrentMap<MetricKey, AverageRecord> map, MetricKey metricKey) {
+        AverageRecord averageRecord = map.get(metricKey);
+        if (averageRecord == null) {
+            AverageRecord newAverageRecord = new AverageRecord();
+            averageRecord = map.putIfAbsent(metricKey, newAverageRecord);
+            if (averageRecord == null) {
+                averageRecord = newAverageRecord;
+            }
+        }
+
+        return averageRecord;
+    }
+
+    private void aggregate(Metric metric) {
+        ConcurrentMap<MetricKey, AverageRecord> timestampMap = getTimestampMap(metric.getTimestamp());
+        MetricKey destinationMetricKey = new MetricKey(
+                metric.getTenant(), metric.getPath(),
+                metric.getRollup(), metric.getPeriod(),
+                metric.getTimestamp());
+        getAverageRecord(timestampMap, destinationMetricKey).addValue(metric.getValue());
+    }
+
+    private void flush() {
+        Collection<Metric> metricsToFlush = new ArrayList<>();
+
+        while(accumulator.size() > 0 && (accumulator.firstKey() < DateTime.now(DateTimeZone.UTC).getMillis() / 1000 - distheneConfiguration.getCarbon().getAggregatorDelay())) {
```

Comment:
well this if condition is extremely huge, can you drop this second statement to a variable ?
I think it will be easier to read

Reply:
fixed